sdk/lib/opentelemetry/sdk/trace/samplers/probability_sampler.rb

```diff
@@ -8,8 +8,87 @@ module OpenTelemetry
+            #   Must be within [0.0, 1.0].
+            # @raise [ArgumentError] if {probability} is out of range
+            # @return [OpenTelemetry::Trace::Samplers::Sampler]
+            def create(probability)
+              raise ArgumentError, 'probability must be in range [0.0, 1.0]' unless (0.0..1.0).include?(probability)
+
+              return OpenTelemetry::Trace::Samplers::NeverSampleSampler.new if probability == 0.0
+              return OpenTelemetry::Trace::Samplers::AlwaysSampleSampler.new if probability == 1.0
+
+              new(probability)
+              # idUpperBound = (long) (probability * Long.MAX_VALUE);
+            end
+          end
+
+          # @api private
+          # The constructor is private and only for use internally by the
+          # class. Users should use the {create} factory method to obtain a
+          # {ProbabilitySampler} instance.
+          #
+          # @param [Numeric] probability The desired probability of sampling.
+          #   Must be within [0.0, 1.0].
+          # @return [ProbabilitySampler]
+          def initialize(probability)
+            @probability = probability
+          end
+
+          SAMPLE_DECISION = OpenTelemetry::Trace::Samplers::Decision.new(decision: true)
+          DONT_SAMPLE_DECISION = OpenTelemetry::Trace::Samplers::Decision.new(decision: false)
+
+          private_constant(:SAMPLE_DECISION, :DONT_SAMPLE_DECISION)
+
+          # Returns the sampling {Decision} for a {Span} to be created
+          #
+          # @param [SpanContext] span_context The {SpanContext} of a parent span,
+          #   typically extracted from the wire. Can be nil for a root span.
+          # @param [Boolean] extracted_context True if span_context was extracted
+          #   from the wire. Can be nil for a root span.
+          # @param [Integer] trace_id The trace_id of the {Span} to be created
+          # @param [Integer] span_id The span_id of the {Span} to be created
+          # @param [String] span_name Name of the {Span} to be created
+          # @param [Enumerable<Link>] links A collection of links to be associated
+          #   with the {Span} to be created. Can be nil.
+          # @return [Decision] The sampling decision
+          def decision(span_context: nil,
+                       extracted_context: nil,
+                       trace_id:,
+                       span_id:,
+                       span_name:,
+                       links: nil)
+            super
+
+            # If the parent is sampled keep the sampling decision.
+            if span_context&.trace_flags&.sampled?
+              SAMPLE_DECISION
+            elsif links&.any? { |link| link.context.trace_flags.sampled? }
+              # If any parent link is sampled keep the sampling decision.
+              SAMPLE_DECISION
+            else
+              # TODO: what should we assume about the distribution of trace_id
+              # bits? How best to compare against @probability?
```

Comment:
Need to figure out the "best" way to compute the sampling decision. I don't think the Java implementation is ideal here. On the other hand, the distribution of `trace_id.hash` probably isn't great either - we have fixed-length IDs with a limited range of characters. I may be totally overthinking that, though. The harder part of this is the scaled comparison - we don't know the range of `#hash` values. In practice, it is truncated to a `Fixnum`, but that detail, and the range of `Fixnum` is obscured in MRI 2.5+.

The OpenTelemetry Service has a probability sampler that computes the murmur3 hash of the trace ID. A Ruby implementation of that won't be efficient on MRI, and we don't want to introduce C extensions in the SDK.

Reply:
It might not be necessary to use the trace_id when computing the probability at all. The javascript implementation basis its decision off `Math.random()`, see: https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-core/src/trace/sampler/ProbabilitySampler.ts

Reply:
👍 that's exactly what we do in our internal tracing gem at Shopify. Happy to stick with that.